### DISCUSSION TITLE
What is the best way to print to a Windows print server?

### DISCUSSION THREAD

**Discussion by xiaowen581** [OpenPrinting/cups#976] [Q&A] [answered]
Windows print server seems to only support IPP 1.0, so the only way to set up the print queue is either "raw" or use the appropriate PPD.

Can the PPD/drivers provided in the link be used to set up the printer shared by the Windows print server? 
https://openprinting.org/drivers

I observed that the document-format-supported for the printers shared by the Windows print server is only application/octet-stream, so I cannot determine if using these PPD/drivers is the correct choice.

Thanks!

**Comment by zdohnal** [comment]
Hi,

AFAIK application/octet-stream is document format for raw file printing, so theoretically it say it supports all formats, but in practice it supports only formats the printer driver on the server can convert into format which is required by the printer (this is again defined by driver on Windows).
Theoretically you can use the OpenPrinting drivers for Linux, but you have to count on Windows driver being able to detect the file was already filtered and does not apply options again (which is sometimes not the case).

What I usually see when there is a Windows server sharing printers is Samba, or IPP+Kerberos, but both raw. You can try the PPD drivers, but it can happen the file from client won't be recognized as already preprocessed and it can lead to errors.

The current Windows servers (based on Win 10 or 11) should be able to use IPP 2.0 and OAuth, and we in CUPS are currently working on OAuth support, which would cover Windows print servers under driverless cloak.

**Reply by michaelrsweet** [reply, marked as answer]
> AFAIK application/octet-stream is document format for raw file printing, so theoretically it say it supports all formats, ...

Actually IPP defines it as "auto-detect", which is why CUPS added its own "raw" MIME media type (application/vnd.cups-raw) to signal that no additional processing should be performed.

Current Windows Server supports Microsoft's "Universal Print Service" architecture which is based on IPP/2.0 and IPP Everywhere. For older Windows Server OS's you are better off using Samba on the Linux/Unix/macOS side with the proper driver.
